### PR TITLE
[LLK] Add bfp2_b support to test infra

### PR DESCRIPTION
**Pull request by halghTT**
### PR Category
Feature

### Summary
Adds `Bfp2_b` (1-bit mantissa block-float) support to the tt-llk Python test infrastructure.

**Format utilities (`bfp_format_utils.py`)**
- `bfp2b_to_float16b`: round-trip quantization using shared `_bfp_prepare_flat` / `_finalize_bfp_quantized` helpers
- Common setup extracted into `_bfp_prepare_flat` (shared by BFP2/4/8); per-format constants unified into module-level `BFP_BLOCK = 16`

**Pack/unpack (`pack.py`, `unpack.py`)**
- `float_to_bfp2_block` / `pack_bfp2_b`: 2-bit packing (4 datums/byte)
- Common BFP scaffolding factored into `_bfp_prepare_blocks` + `_bfp_collect_blocks`, shared across BFP8/4/2
- `unpack_bfp2_b` unified with BFP4/8 into a single generalised `_unpack_bfp_b` helper

**Stimuli generation (`stimuli_generator/`)**
- `_default_bfp2b_face`, `DataFormat.Bfp2_b` in `_get_dtype_for_format`, `_generate_source_tensor` quantization paths, and `default_spec_for_format` — integrated into the refactored `stimuli_generator/` package (after main converted the single file into a `generator.py` / `utils.py` / `spec.py` / `strategies/` package)

**Golden models (`golden_generators.py`)**
- `Bfp2_b` added to `quantize_input_to_unpack_format` and all quantization dispatch paths
- `_src_reg_format` fidelity helper includes `Bfp2_b` (unpacks to `Float16_b` in src regs)
- `_quantize_reduce_input` / `_quantize_reduce_output` updated for `Bfp2_b`

**Test comparison (`utils.py`)**
- `Bfp2_b` tolerance: `atol=0.5, rtol=0.4`, PCC threshold `0.90`
- `_bfp_block_aware_compare` parameterised on `mantissa_bits`; called with `mantissa_bits=1, max_ulp_diff=1` for `Bfp2_b` to prevent sign-flip masking

**Test (`test_eltwise_unary_datacopy.py`)**
- Shared `_run_unary_datacopy_test` helper; `test_unary_datacopy_sub_byte_bfp` covers `Bfp2_b` and `Bfp4_b` with `quantize_golden_input=True`

### Notes for reviewers
- BFP2 operates directly on float32 bit patterns for the exponent comparison; no mantissa rounding step is needed since there are no mantissa bits to round
- `max_ulp_diff=1` (not 2) for BFP2 prevents `+scale` vs `−scale` from being treated as within tolerance
- Bfp4_b tolerance tightened to `atol=0.125`, PCC `0.97` to align with upstream

### CODE REVIEW

File: tt_metal/tt-llk/tests/python_tests/helpers/bfp_format_utils.py [hunk 1]
```diff
@@ -118,3 +118,64 @@ def bfp4b_to_float16b(operand: torch.Tensor, dimensions=None) -> torch.Tensor:
...
+
+        { 0, +1 * 2^(shared_exp - 127), -1 * 2^(shared_exp - 127) }
+
+    Args:
+        operand: Input tensor (any shape).
+        dimensions: If provided, untilize the result back to these dimensions.
+
+    Returns:
+        Quantized bfloat16 tensor (same number of elements as input).
+    """
```
Comment (rtawfik01) [inline tt_metal/tt-llk/tests/python_tests/helpers/bfp_format_utils.py:141]: The first 6 lines here are the same for BFP2/4/8, I think we should extract these and reuse them for all BFP formats

File: tt_metal/tt-llk/tests/python_tests/helpers/pack.py [hunk 2]
```diff
@@ -257,6 +257,87 @@ def pack_bfp4_b(tensor, block_size=16, num_faces=4, face_r_dim=16):
...
+
+    raw_bytes = struct.pack(f"<{n}f", *(float(v) for v in block))
+    all_bits = struct.unpack(f"<{n}I", raw_bytes)
+
+    signs = [0] * n
+    exponents = [0] * n
+    mantissas = [0] * n
+    shared_exponent = 0
+
+    for i, bits in enumerate(all_bits):
```
Comment (rtawfik01) [inline tt_metal/tt-llk/tests/python_tests/helpers/pack.py]: Same point as earlier, this section can be abstracted to a helper function for all BFP formats, in general the helpers for BFP formats should be consolidated since the only difference is the number of mantissa bits

**Comment by halghTT** [comment]
@copilot resolve the merge conflicts in this pull request

File: tt_metal/tt-llk/tests/python_tests/test_eltwise_unary_datacopy.py [hunk 3]
```diff
@@ -186,114 +191,62 @@ def test_unary_datacopy(
     assert passed_test(golden_tensor, res_tensor, formats.output_format)
 
 
+@parametrize(
+    formats=input_output_formats(
+        [
+            DataFormat.Float32,
+            DataFormat.Float16,
+            DataFormat.Float16_b,
+            DataFormat.Bfp8_b,
```
Comment (rtawfik01) [inline tt_metal/tt-llk/tests/python_tests/test_eltwise_unary_datacopy.py:200]: Bfp2_b should be part of this test list to validate it? This test list just does a L1-to-L1 datacopy, so it will be perfect for verifying that Bfp2_b is correctly unpack to Fp16_b and packed again to Bfp2_b

Comment (halghTT) [reply tt_metal/tt-llk/tests/python_tests/test_eltwise_unary_datacopy.py:200]: It is part of the test - see below (line 230)

File: tt_metal/tt-llk/tests/python_tests/helpers/bfp_format_utils.py [hunk 4]
```diff
@@ -64,17 +64,26 @@ def _finalize_bfp_quantized(
     return out
 
 
-def bfp8b_to_float16b(operand: torch.Tensor, dimensions=None) -> torch.Tensor:
-    """
-    Simulate BFP8_b pack/unpack round-trip quantization.
+def _quantize_bfp_b(
```
Comment (rtawfik01) [inline tt_metal/tt-llk/tests/python_tests/helpers/bfp_format_utils.py:67]: nice cleanup!

File: tt_metal/tt-llk/tests/python_tests/helpers/unpack.py [hunk 5]
```diff
@@ -219,6 +219,66 @@ def unpack_bfp4_b(bfp4_block, sfpu=False, num_faces=4, face_r_dim=16):
     return torch.tensor(bfloat16_values, dtype=torch.bfloat16)
 
 
+def bfp2_to_float_block(exponent, bfp2_mantissas, unpacked_bfp2):
+    bfloat16_values = []
+    exp_adj = exponent - 127
+    scale = 2.0**exp_adj
+
```
Comment (rtawfik01) [inline tt_metal/tt-llk/tests/python_tests/helpers/unpack.py]: I think you are missing a check for inf:

```
if mag == 0:
    value = -float("inf") if (mantissa & 0x2) else 0.0
    bfloat16_values.append(value)
    unpacked_bfp2[(exp_adj, mantissa)] = value
    continue
```

Comment (rtawfik01) [reply tt_metal/tt-llk/tests/python_tests/helpers/unpack.py]: The bfp4 and bfp8 path has it

File: tt_metal/tt-llk/tests/python_tests/helpers/unpack.py [hunk 6]
```diff
@@ -219,6 +219,66 @@ def unpack_bfp4_b(bfp4_block, sfpu=False, num_faces=4, face_r_dim=16):
...
+    for mantissa in bfp2_mantissas:
+        mag = mantissa & 0x1
+        if mag == 0:
+            bfloat16_values.append(0.0)
+            unpacked_bfp2[(exp_adj, mantissa)] = 0.0
```
Comment (rtawfik01) [inline tt_metal/tt-llk/tests/python_tests/helpers/unpack.py]: I think this should be:
```
if mag == 0:
    bfloat16_values.append(0.0)
    continue
```

File: tt_metal/tt-llk/tests/python_tests/helpers/unpack.py [hunk 7]
```diff
@@ -219,6 +219,66 @@ def unpack_bfp4_b(bfp4_block, sfpu=False, num_faces=4, face_r_dim=16):
     return torch.tensor(bfloat16_values, dtype=torch.bfloat16)
 
 
+def bfp2_to_float_block(exponent, bfp2_mantissas, unpacked_bfp2):
```
Comment (rtawfik01) [inline tt_metal/tt-llk/tests/python_tests/helpers/unpack.py]: Is it possible to also make this function generalizable for bfp2/4/8_b? Otherwise all of the bfp24/8_to_float functions look almost the same

**Review by rtawfik01 — APPROVED** [review]

**Comment by rtawfik01** [comment]
@ldjurovicTT @uvelimirovicTT please take a look

**Comment by halghTT** [comment]
@copilot resolve the merge conflicts in this pull request

**Comment by halghTT** [comment]
@copilot resolve the merge conflicts in this pull request
